### PR TITLE
moabs: add more options

### PR DESCRIPTION
**Pull request by lijinbio**
FOR CONTRIBUTOR:
* [x] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [x] - License permits unrestricted use (educational + commercial)
* [ ] - This PR adds a new tool or tool collection
* [x] - This PR updates an existing tool or tool collection
* [ ] - This PR does something else (explain below)

### CODE REVIEW

File: tools/moabs/moabs.xml [hunk 1]
```diff
@@ -147,44 +187,98 @@
...
                 </when>
             </conditional>
-            <param argument="-n" type="boolean" truevalue="1" falsevalue="0" checked="true" label="Mapping to four strands?" help="Yes: map SE or PE reads to all 4 strands, i.e. ++, +-, -+, --; No: only map to 2 forward strands, i.e. BSW(++) and BSC(-+)"/>
-            <param argument="-r" type="select" label="How to report repeat hits" help="0=none(unique hit/pair); 1=random one; 2=all(slow)">
+            <param argument="-w" type="integer" value="1000" min="0" max="1000" label="Maximum number of equal best hits to count" help="Maximum number of equal best hits to count. When multiple mapping occurs for a read, it should control the number of records to report. Default: 1000."/>
+            <param argument="-D" type="text" label="Restriction enzyme digestion sites for RRBS mode" help="For RRBS data, this option activates RRBS mapping mode and set restriction enzyme digestion sites. Digestion position marked by '-', example: -D C-CGG for MspI digestion.  default: none (WGBS mode)."/>
```
Comment (bgruening) [inline tools/moabs/moabs.xml:191]: Can you please add a validator and restrict the user input.

File: tools/moabs/moabs.xml [hunk 2]
```diff
@@ -147,44 +187,98 @@
...
+            <param argument="-q" type="integer" value="0" min="0" max="40" label="Quality threshold in trimming" help="The quality threshold to trim read bases. To obtain an accurate mapping, low-quality bases should be skipped beforehand. Min=0, max=40. Default=0 (no trim)."/>
+            <param argument="-z" type="select" label="Base quality" help="Base quality for sequencing reads, Illumina or Sanger.">
+                <option value="33" selected="true">Sanger</option>
+                <option value="64">Illumina</option>
+            </param>
+            <param argument="-f" type="integer" value="5" min="0" label="Maximum number of Ns in a read to filter out" help="To filter out low-quality reads containing >n Ns. Default=5."/>
+            <param argument="-A" type="text" label="3' adapter sequence to trim" help="To trim 3' adapter sequence. Default: none (no trim)."/>
```
Comment (bgruening) [inline tools/moabs/moabs.xml:200]: Please add validator.

File: tools/moabs/moabs.xml [hunk 3]
```diff
@@ -147,44 +187,98 @@
...
+            <param argument="--minCredibleDif" type="float" value="0.2" min="0" max="1" label="Minimal credible methylation difference for DMC calling" help="Minimal CDIF for DMC calling. Default=`0.2`."/>
+            <param argument="--minDmcsInDmr" type="integer" value="3" min="1" label="Minimum number of DMCs in a DMR" help="Minimum number of DMCs in a DMR. Default=3."/>
+            <param argument="--maxDistConsDmcs" type="integer" value="300" min="1" label="Maximum distance between two consective DMCs for a DMR" help="Maximum distance between two consective DMCs for a DMR. Default=300."/>
         </section>
         <param name="output_selector" type="select" multiple="true" optional="true" label="Select output files" help="">
-            <option value="1"> dmr_M3_g1.G.bed_vs_g2.G.bed.txt.dmr </option>
-            <option value="2"> comp.g1.vs.g2.txt </option>
-            <option value="3"> BAM files </option>
-            <option value="4"> Methylation calling BED files </option>
+            <option value="1" selected="true"> dmc_M2_g1.G.bed_vs_g2.G.bed.txt </option>
```
Comment (bgruening) [inline tools/moabs/moabs.xml:259]: can we remove the first space?

**Comment by lijinbio** [comment]
Hi @bgruening, thank you for your comments. I will address them and update the PR accordingly.

**Comment by lijinbio** [comment]
> @lijinbio this is awesome. Can you please bump the version number of the tool.

As for the version number of the tool, may I set 1.3.4.6.1 as the version id in the header? although the requirement of the MOABS version is not changed, still depending on MOABS 1.3.4.6.

**Comment by bgruening** [comment]
@lijinbio please see here: https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html#tool-versions

File: tools/moabs/moabs.xml (hunk 3 above)
Comment (lijinbio) [reply tools/moabs/moabs.xml]: Yes, the first and last spaces are removed.

File: tools/moabs/moabs.xml (hunk 2 above)
Comment (lijinbio) [reply tools/moabs/moabs.xml]: Yes, a validator is added for the text inputs.

File: tools/moabs/moabs.xml (hunk 1 above)
Comment (lijinbio) [reply tools/moabs/moabs.xml]: Yes, a validator is added for user input.

**Comment by lijinbio** [comment]
@bgruening Hi Björn, I have addressed your comments, and the commit has passed the lint checks. Can you help review my PR again? Thank you.

**Comment by bgruening** [comment]
@lijinbio sorry, unfortunaltey, the test are failling with:

Files /tmp/tmpyquix209SE_g1_r1.bam=96198b but /tmp/tmp8n3r_2vfSE_g1_r1.bam=111199b - compare by size (delta=10000) failed


I know its green sorry, we changed the testing recently and still need to iron out something.

The error indicates that your BAM files do have changed. BAM files by the way can be compared with normal diff, as Galaxy converts them to SAM and then compares the SAM files. So please remove the sim_size in your BAM test outputs.

**Comment by lijinbio** [comment]
> @lijinbio sorry, unfortunaltey, the test are failling with:
> 
> Files /tmp/tmpyquix209SE_g1_r1.bam=96198b but /tmp/tmp8n3r_2vfSE_g1_r1.bam=111199b - compare by size (delta=10000) failed
> 
> I know its green sorry, we changed the testing recently and still need to iron out something.
> 
> The error indicates that your BAM files do have changed. BAM files by the way can be compared with normal diff, as Galaxy converts them to SAM and then compares the SAM files. So please remove the sim_size in your BAM test outputs.

@bgruening Hi Björn, yes, I realized I have added some alignment options to BSMAP, and the generated BAM may change w.r.t. the command line in the header. May I ask how could I run the test locally so that I can generate the exact BAM files by mimicking the web interface use? Thank you.

At the same time, I will change the test command as below. Do they follow the convention to test a BAM file?

```
            <output_collection name="output_collection_bam" count="4">
                <element name="g1_r1.bam" file="SE_g1_r1.bam" ftype="bam"/>
                <element name="g1_r2.bam" file="SE_g1_r2.bam" ftype="bam"/>
                <element name="g2_r1.bam" file="SE_g2_r1.bam" ftype="bam"/>
                <element name="g2_r2.bam" file="SE_g2_r2.bam" ftype="bam"/>
            </output_collection>
```

**Comment by lijinbio** [comment]
@bgruening For example, the BAM header looks like below.

```
@HD	VN:1.0	SO:coordinate
@SQ	SN:chr11	LN:30000
@PG	ID:BSMAP	VN:2.87	CL:"bsmap -a /tmp/tmpmjC4hb/files/2/2/a/dataset_22a0b1fd-69cf-4fbe-a4bd-f31d0aa36ba9.dat -o g1_r1.bam -n 1 -d /tmp/tmpmjC4hb/files/f/e/8/dataset_fe8e1e6e-fc58-4ee5-8b8a-8663fb954d17.dat -r 0 -R -p 1"
```

Yet it will definitely change when running in the CI check. May I ask how could I make sure my input BAM will match the test BAMs? Thank you.

**Comment by lijinbio** [comment]
> @lijinbio sorry, unfortunaltey, the test are failling with:
> 
> Files /tmp/tmpyquix209SE_g1_r1.bam=96198b but /tmp/tmp8n3r_2vfSE_g1_r1.bam=111199b - compare by size (delta=10000) failed
> 
> I know its green sorry, we changed the testing recently and still need to iron out something.
> 
> The error indicates that your BAM files do have changed. BAM files by the way can be compared with normal diff, as Galaxy converts them to SAM and then compares the SAM files. So please remove the sim_size in your BAM test outputs.

@bgruening Hi Björn, I have removed the sim_size but used lines_diff for the BAM test. Can you help review my PR again? Thank you.

**Comment by bgruening** [comment]
@lijinbio thanks a lot! This looks great!

**Comment by lijinbio** [comment]
> @lijinbio thanks a lot! This looks great!

@bgruening Hi Björn, many thanks for reviewing and merging my PR.

I found that the interface has not been upgraded yet at https://usegalaxy.eu/root?tool_id=moabs, may I ask if I need to submit a PR to change the revision number in `tools_iuc.yaml.lock` in the usegalaxy-eu-tools repository? so that the interface will update accordingly. Thank you for your help.

**Comment by lijinbio** [comment]
> > @lijinbio thanks a lot! This looks great!
> 
> @bgruening Hi Björn, many thanks for reviewing and merging my PR.
> 
> I found that the interface has not been upgraded yet at https://usegalaxy.eu/root?tool_id=moabs, may I ask if I need to submit a PR to change the revision number in `tools_iuc.yaml.lock` in the usegalaxy-eu-tools repository? so that the interface will update accordingly. Thank you for your help.

@bgruening Hi Björn, I have submitted a PR to add the revision of MOABS at UseGalaxy.eu (
https://github.com/usegalaxy-eu/usegalaxy-eu-tools/pull/291), can you help review my PR there? Thank you.

**Comment by bgruening** [comment]
Awesome, thanks!

**Pull request merged**